File: 4.cpp
#include <bits/stdc++.h>
using namespace std;

vector<pair<int, int>> directions = {
    {0, 1}, {1, 0}, {1, 1}, {1, -1}, // right, down, down-right, down-left
    {0, -1}, {-1, 0}, {-1, -1}, {-1, 1} // left, up, up-left, up-right
};

bool is_valid(int x, int y, int n, int m) {
    return x >= 0 && x < n && y >= 0 && y < m;
}

bool search_word(vector<vector<char>>& grid, int x, int y, string word) {
    int n = grid.size();
    int m = grid[0].size();
    for (auto dir : directions) {
        int dx = dir.first, dy = dir.second;
        int k;
        for (k = 0; k < word.size(); k++) {
            int nx = x + k * dx;
            int ny = y + k * dy;
            if (!is_valid(nx, ny, n, m) || grid[nx][ny] != word[k]) {
                break;
            }
        }
        if (k == word.size()) {
            return true;
        }
    }
    return false;
}

int count_occurrences(vector<vector<char>>& grid, string word) {
    int n = grid.size();
    int m = grid[0].size();
    int count = 0;
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < m; j++) {
            if (grid[i][j] == word[0] && search_word(grid, i, j, word)) {
                count++;
            }
        }
    }
    return count;
}

int main() {

    vector<vector<char>> grid;
    for (int i = 0; i < n; i++) {
        vector<char> row;
        for (int j = 0; j < m; j++) {
            char tmp;cin>>tmp;

            row.push_back(tmp);
        }
        grid.push_back(row);
    }

    string word = "XMAS";
    int result = count_occurrences(grid, word);
    cout << result << endl;

    return 0;
}

File: 4.py
def find_word(grid, word):
    rows, cols = len(grid), len(grid[0])
    word_len = len(word)
    count = 0

    def check_direction(x, y, dx, dy):
        for i in range(word_len):
            if not (0 <= x < rows and 0 <= y < cols) or grid[x][y] != word[i]:
                return False
            x += dx
            y += dy
        return True

    directions = [
        (0, 1), (1, 0), (1, 1), (1, -1),  # right, down, down-right, down-left
        (0, -1), (-1, 0), (-1, -1), (-1, 1)  # left, up, up-left, up-right
    ]

    for r in range(rows):
        for c in range(cols):
            for dx, dy in directions:
                if check_direction(r, c, dx, dy):
                    count += 1

    return count

# Example usage
grid = []
while True:
    a = input()
    if a == "dd":
        break
    grid.append(list(a))

word = "XMAS"
# print(find_word(grid, word))  # Output: 18
count = 0
for i in range(len(grid)-2):
    for j in range(len(grid[i])-2):
            if( (grid[i][j] == 'M' and grid[i+1][j+1] == 'A' and grid[i+2][j+2] == 'S')\
                or (grid[i][j] == 'S' and grid[i+1][j+1] == 'A' and grid[i+2][j+2] == 'M'))\
                and(( grid[i+2][j] == 'M' and grid[i][j+2] == 'S') or (grid[i+2][j] == 'S' and grid[i][j+2] == 'M')):
                    count += 1
print(count)